### PR TITLE
Payments/consumer side balance tracking

### LINKED ISSUE #716: Adjust provider/consumer balance after receiving/sending a promise
Currently, once the promise is sent, not much is done with it. We need to make sure that the balance accounts for the increase in payment for the provider. The consumer will probably want to keep track of this as well.

### LINKED ISSUE #717: Consumer-side balance tracking
At the moment, the consumer does not keep balance on his side. Basically needs to do a similar thing to the provider side balance tracking.

We are also passing a bad param to extendPromise method. This needs to be passed in correctly, once we have a proper balance keeper.

See:
https://github.com/mysteriumnetwork/node/pull/692/files/1fc5cf0315198d2385255252410d8ebf98631cbe#r254980944

Adding the same comment below as well:

We should have a logic there. Current dump balance addition should be pointed out and separate task should exists to implement this logic. If we get balance not enough for next charge period we should extend promise with enough credits just to cover future expected communication. Consumers confidence (reliance on provider) should reflected on this logic. Probably we should have some external parameter to control this from application settings.

### PR DESCRIPTION
**Pull request by vkuznecovas**
marking this as a draft since I need the previous 2 to get merged so i can rebase this properly and merge. Closes #717 and Closes #716

### CODE REVIEW

File: core/connection/manager.go [hunk 1]
```diff
@@ -21,13 +21,16 @@ import (
...
+	"time"
 
 	log "github.com/cihub/seelog"
 	"github.com/mysteriumnetwork/node/communication"
 	"github.com/mysteriumnetwork/node/consumer"
 	"github.com/mysteriumnetwork/node/firewall"
 	"github.com/mysteriumnetwork/node/identity"
 	"github.com/mysteriumnetwork/node/market"
+	"github.com/mysteriumnetwork/node/money"
+	"github.com/mysteriumnetwork/node/services/openvpn/discovery/dto"
```
Comment (soffokl) [inline core/connection/manager.go]: Again I'm worried that we are using openvpn specific `dto.PaymentPerTime`.

Comment (vkuznecovas) [reply core/connection/manager.go]: I'll create a separate ticket to move it to somewhere more appropriate.

Comment (vkuznecovas) [reply core/connection/manager.go]: https://github.com/mysteriumnetwork/node/issues/761

File: session/payment/session_balance.go [hunk 2]
```diff
@@ -136,43 +134,20 @@ func (sb *SessionBalance) Start() error {
 }
 
 func (sb *SessionBalance) loadInitialPromiseState() (promise.StoredPromise, error) {
-	lastPromise, err := sb.promiseStorage.GetLastPromise(sb.issuerID)
+	lastPromise, err := sb.promiseStorage.FindPromiseForConsumer(sb.issuerID, sb.consumerID)
 	if err != nil {
-		if err.Error() == errBoltNotFound.Error() {
+		if err.Error() == errBoltNotFound.Error() || err.Error() == errNoPromiseForConsumer.Error() {
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply session/payment/session_balance.go]: Two errors with the same message are not equal. See: https://play.golang.org/p/rxwh2j6okm2

I could get rid of this issue by having a direct import of the bolt package, but I don't want the balance to know anything about the underlying DB. Therefore I have to hide it by declaring my own variable. I could just store it in a string though, not sure if that would be more readable though. I feel like mirroring the same error here is an OK solution to this problem.

File: session/promise/storage.go [hunk 3]
```diff
@@ -193,14 +199,36 @@ func (s *Storage) GetAllPromisesFromIssuer(issuerID identity.Identity) ([]Stored
...
-		return &PaymentInfo{}
+		if err == errNoPromiseForConsumer {
+			lastPromise, lastPromiseErr := s.getLastPromise(issuerID)
+			if lastPromiseErr != nil {
+				log.Trace(promiseLogPrefix, "could not load promise info, defaulting to zero payment info ", lastPromiseErr)
+				return defaultPaymentInfo
+			}
+
+			newID := lastPromise.SequenceID + 1
+			log.Trace(promiseLogPrefix, "found payments for other consumers, will issue promiseID ", newID)
```
Comment (Waldz) [inline session/promise/storage.go]: Quite unreadable here, can we move sequence increment logic to storage level and lets have proper locks for that

File: session/promise/storage_test.go [hunk 4]
```diff
@@ -195,17 +201,53 @@ func Test_LoadPaymentInfo_LoadsProperInfo(t *testing.T) {
...
 	assert.Equal(t, uint64(300), pi.LastPromise.Amount)
 	assert.Equal(t, uint64(200), pi.FreeCredit)
 }
 
+func Test_LoadPaymentInfo_WithPreviousConsumersIssuesNewID(t *testing.T) {
+	newMock := map[string][]StoredPromise{
+		getBucketNameFromIssuer(issuerID): {
+			{
+				SequenceID: 10,
+				AddedAt:    timeMock,
```
Comment (Waldz) [inline session/promise/storage_test.go]: Where is `consumerID` in mocked DB state?

File: session/payment/factory/payments.go [hunk 5]
```diff
@@ -51,21 +57,30 @@ func noopPaymentIssuerFactory(initialState promise.State,
...
 		messageChan chan balance.Message,
 		dialog communication.Dialog,
 		consumer, provider identity.Identity) (connection.PaymentIssuer, error) {
 
 		bl := balance.NewListener(messageChan)
 		ps := promise.NewSender(dialog)
 		issuer := issuers.NewLocalIssuer(signerFactory(consumer))
-		tracker := promise.NewConsumerTracker(initialState, consumer, provider, issuer)
-		payments := payment.NewSessionPayments(messageChan, ps, tracker)
+		tracker := promise.NewConsumerTracker(promise.State{
```
Comment (Waldz) [inline session/payment/factory/payments.go]: Looks like too much glue logic in factory, such fat factories demands to be tested.
Or can we move them out somehow?

File: session/promise/storage_test.go (hunk 4 above)
Comment (vkuznecovas) [reply session/promise/storage_test.go]: the defaults act as a separate consumer, but i've made it more explicit anwyay.

File: session/payment/factory/payments.go (hunk 5 above)
Comment (vkuznecovas) [reply session/payment/factory/payments.go]: I've added a test for the only piece of logic we had there.

File: session/payment/factory/payments.go [hunk 6]
```diff
@@ -51,22 +57,37 @@ func noopPaymentIssuerFactory(initialState promise.State,
...
 		dialog communication.Dialog,
 		consumer, provider identity.Identity) (connection.PaymentIssuer, error) {
 
 		bl := balance.NewListener(messageChan)
 		ps := promise.NewSender(dialog)
 		issuer := issuers.NewLocalIssuer(signerFactory(consumer))
-		tracker := promise.NewConsumerTracker(initialState, consumer, provider, issuer)
-		payments := payment.NewSessionPayments(messageChan, ps, tracker)
+
+		promiseState := mapInitialStateToPromiseState(initialState)
```
Comment (Waldz) [inline session/payment/factory/payments.go]: Do this initialisation inside `promise.NewConsumerTracker()`

**Comment by Waldz** [comment]
What's the status here?

**Comment by vkuznecovas** [comment]
We should merge it. That's the status :D

**Pull request merged**